### ISSUE TITLE
Panic on Unwrap in sharedparley::layout

### ISSUE THREAD

**Issue by qarmin** [slint-ui/slint#10336]
### Feature Description

Recently, one of the users who was running the application inside a VM reported a panic in my app.
```
18:26:27.495 [INFO] czkawka_core::common::logger: Logging to file "/home/yanganto/.cache/czkawka/krokiet.log" and terminal
18:26:27.499 [INFO] czkawka_core::common::logger: Krokiet version: 10.0.0(<unknown> U <unknown>), release mode, rust 1.90.0 (2025-09-14), os NixOS 25.11.0 (x86_64 64-bit), 32 cpu/threads, features(0): [], app cpu version: x86-64-v1 (SSE2), os cpu version: x86-64-v4 (AVX-512), libc glibc [build + runtime (glibc 2.40 | musl <not detected>)]
18:26:27.499 [INFO] czkawka_core::common::config_cache_path: Config folder set to "/home/yanganto/.config/krokiet" and cache folder set to "/home/yanganto/.cache/czkawka"
18:26:27.499 [INFO] krokiet: Krokiet features(2): [winit_femtovg, winit_software]
18:26:27.609 [ERROR] log_panics: thread 'main' panicked at 'called `Option::unwrap()` on a `None` value': /nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/i-slint-core-1.14.1/textlayout/sharedparley.rs:229
   0: log_panics::Config::install_panic_hook::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys::backtrace::__rust_end_short_backtrace
   4: __rustc::rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::panicking::panic
   7: core::option::unwrap_failed
   8: i_slint_core::textlayout::sharedparley::layout
   9: i_slint_core::textlayout::sharedparley::draw_text
  10: <i_slint_renderer_femtovg::itemrenderer::GLItemRenderer<R> as i_slint_core::item_rendering::ItemRenderer>::draw_text
  11: i_slint_core::items::ComplexTextVTable::render
  12: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  13: i_slint_core::item_tree::visit_item_tree
```

So the crash clearly happened in the text layouting code path. When looking at the source:

https://github.com/slint-ui/slint/blob/ce50ea806a9a1d512d30acab6f99c8a1d511505f/internal/core/textlayout/sharedparley.rs#L229

there is a plain, unconditional unwrap without any explanation why this cannot panic, instead of proper error handling.

This is not the first time I’ve seen random panics in Slint, so I think the entire repository (except maybe auto-generated code) should use the following clippy flag:
```
#![warn(clippy::unwrap_used)]
```
and replace all unwrap calls with proper error handling or expect calls that include a meaningful message, which would help with diagnosing the root cause of such errors.

**Comment by ogoffart** [comment]
Agree that we should attempt to make sure that the code never panics. But there are valid usage of unwrap.
In this case, it is clearly a bug that should be fixed. (Would you happen to have a testcase?)

**Comment by thomas725** [comment]
Hello @ogoffart ! I'm the user trying to run the new slint based gui of "czkawka" = "krokiet" on my NixOS machine.

We've managed to setup the required dependencies wayland, libxkbcommon & libGL in a way to get them visible for slint (or it's dependencies), but now we're crashing with the shown unwrap on none panic. For details see: https://github.com/qarmin/czkawka/issues/1720

For the 3 listed dependencies it was easy to see what was missing:
* `winit-0.30.12/src/platform_impl/linux/wayland/event_loop/mod.rs:89: The wayland library could not be loaded`
* `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: XKBNotFound': winit-0.30.12/src/platform_impl/linux/wayland/seat/keyboard/mod.rs:300`
* `thread 'main' panicked at 'Failed to run app :(: Error creating OpenGL display (Ok(DisplayHandle(Wayland(WaylandDisplayHandle { display: 0x5599b3ba0760 })))) with glutin: not found': krokiet/src/main.rs:114`

But that 4th error sadly doesn't show any hint on what is going on (at least not for people who aren't familiar with the code..)
```
[ERROR] log_panics: thread 'main' panicked at 'called `Option::unwrap()` on a `None` value': /build/czkawka-10.0.0+-vendor/i-slint-core-1.14.1/textlayout/sharedparley.rs:229
   0: log_panics::Config::install_panic_hook::{{closure}}
   1: std::panicking::panic_with_hook
   2: std::panicking::panic_handler::{{closure}}
   3: std::sys::backtrace::__rust_end_short_backtrace
   4: __rustc::rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::panicking::panic
   7: core::option::unwrap_failed
   8: i_slint_core::textlayout::sharedparley::layout
   9: i_slint_core::textlayout::sharedparley::draw_text
  10: <i_slint_renderer_femtovg::itemrenderer::GLItemRenderer<R> as i_slint_core::item_rendering::ItemRenderer>::draw_text
  11: i_slint_core::items::ComplexTextVTable::render
  12: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  13: i_slint_core::item_tree::visit_item_tree
  14: <krokiet::slint_generatedMainWindow::InnerComponent_empty_87 as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  15: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  16: i_slint_core::item_tree::visit_item_tree
  17: <krokiet::slint_generatedMainWindow::InnerComponent_empty_87 as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  18: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  19: i_slint_core::item_tree::visit_item_tree
  20: <krokiet::slint_generatedMainWindow::InnerComponent_empty_87 as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  21: i_slint_core::model::Repeater<C>::visit
  22: krokiet::slint_generatedMainWindow::InnerSelectableTableView_root_81::visit_dynamic_children
  23: i_slint_core::item_tree::visit_item_tree::{{closure}}
  24: i_slint_core::item_tree::visit_item_tree
  25: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  26: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  27: i_slint_core::item_tree::visit_item_tree::{{closure}}
  28: i_slint_core::item_tree::visit_item_tree
  29: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  30: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  31: i_slint_core::item_tree::visit_item_tree::{{closure}}
  32: i_slint_core::item_tree::visit_item_tree
  33: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  34: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  35: i_slint_core::item_tree::visit_item_tree::{{closure}}
  36: i_slint_core::item_tree::visit_item_tree
  37: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  38: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  39: i_slint_core::item_tree::visit_item_tree::{{closure}}
  40: i_slint_core::item_tree::visit_item_tree
  41: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  42: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  43: i_slint_core::item_tree::visit_item_tree::{{closure}}
  44: i_slint_core::item_tree::visit_item_tree
  45: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  46: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  47: i_slint_core::item_tree::visit_item_tree::{{closure}}
  48: i_slint_core::item_tree::visit_item_tree
  49: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  50: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  51: i_slint_core::item_tree::visit_item_tree::{{closure}}
  52: i_slint_core::item_tree::visit_item_tree
  53: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  54: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  55: i_slint_core::item_tree::visit_item_tree::{{closure}}
  56: i_slint_core::item_tree::visit_item_tree
  57: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  58: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  59: i_slint_core::item_tree::visit_item_tree::{{closure}}
  60: i_slint_core::item_tree::visit_item_tree
  61: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  62: i_slint_core::item_tree::<impl i_slint_core::item_tree::ItemVisitor_vtable_mod::ItemVisitor for T>::visit_item
  63: i_slint_core::item_tree::visit_item_tree::{{closure}}
  64: i_slint_core::item_tree::visit_item_tree
  65: <krokiet::slint_generatedMainWindow::InnerMainWindow as i_slint_core::item_tree::ItemTree_vtable_mod::ItemTree>::visit_children_item
  66: i_slint_core::item_rendering::render_component_items
  67: i_slint_renderer_femtovg::FemtoVGRenderer<B>::internal_render_with_post_callback::{{closure}}
  68: i_slint_core::properties::PropertyTracker<DirtyHandler>::evaluate_as_dependency_root
  69: i_slint_renderer_femtovg::FemtoVGRenderer<B>::render
  70: <i_slint_backend_winit::renderer::femtovg::GlutinFemtoVGRenderer as i_slint_backend_winit::renderer::WinitCompatibleRenderer>::render
  71: i_slint_backend_winit::winitwindowadapter::WinitWindowAdapter::draw
  72: <i_slint_backend_winit::event_loop::EventLoopState as winit::application::ApplicationHandler<i_slint_backend_winit::SlintEvent>>::window_event
  73: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  74: winit::platform_impl::linux::wayland::event_loop::EventLoop<T>::pump_events
  75: winit::platform_impl::linux::EventLoop<T>::run_on_demand
  76: <i_slint_backend_winit::Backend as i_slint_core::platform::Platform>::run_event_loop
  77: slint::run_event_loop
  78: <krokiet::slint_generatedMainWindow::MainWindow as i_slint_core::api::ComponentHandle>::run
  79: krokiet::main
  80: std::sys::backtrace::__rust_begin_short_backtrace
  81: std::rt::lang_start::{{closure}}
  82: std::rt::lang_start_internal
  83: main
  84: __libc_start_call_main
  85: __libc_start_main_alias_1
  86: _start
```

Do you @ogoffart maybe have a simple hello world slint example that does use the same component that I could try to checkout from some repository and `cargo run` to see if that one would be affected too?

**Comment by yanganto** [comment]
Hi @ogoffart,
If you want my two cents, we can change `unwrap` to `expect` with clear message, then developers will easy to follow on what environment the lib expected to run on.  The valid `unwrap` is nice in test cases, and `expect` is just a quicker way to throw out a error message in the functions of crate.

However, a crate nice to have a more concrete error type and handle all errors in Rust, for example

```rust
use thiserror::Error;

#[derive(Error, Debug)]
pub enum SlintError {
    #[error("Current resource are less than the minimal requirement to run: `{0}`")]
    LackResource(String),
}
```

Slint is building a Rust crate for UI and developers are looking for a better world.  With clear message the bugs and issue will be less, and also we can save time on triaging the bugs.
Thank you.

**Comment by ogoffart** [comment]
@yanganto I agree. In Slint we try to not panic. An `unwrap` or `expect` is actually a bug. We normally don't panic and return error instead. Unless there is no way to return an error in which case we sometime still panic with the best possible error messages if we can't recover.

Regarding the errors from @thomas725 's comment:
 - The first two ones are panic in the winit crate. This is indeed not ideal that winit panics like that.
 - The third one is a panic in krokiet from a Slint error.  (The error message is a bit confusing with the debug of the display, maybe that can be improved in https://github.com/slint-ui/slint/blob/e41a460b135175a570e39c95985ded246e368bfb/internal/backends/winit/renderer/femtovg/glcontext.rs#L95 )
 - The 4th is a bug in Slint and that's why this issue is open.

**Comment by ogoffart** [comment]
Fixed by #10384

**Comment by thomas725** [comment]
Thanks! I managed to get Krokiet working by switching slint from 1.13 to github master, see:
https://github.com/thomas725/czkawka/tree/master

**Comment by tronical** [comment]
Thank you all from the Krokiet project for the excellent bug reports.

Olivier found a fix for the wayland resize and I think we can fix the ellipsis bug also. So fingers crossed the next 1.15 release will be good for you folks :)
